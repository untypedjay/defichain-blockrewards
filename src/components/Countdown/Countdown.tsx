import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import { cardStyles } from '../../styles/CardStyle';
import { CountdownElement } from './index';

interface Props {
  date: Date;
}

const StyledCountdown = styled.div`
  ${cardStyles};
  display: flex;
  width: 100%;
  justify-content: space-between;
`;

export default function Countdown({ date }: Props) {
  const calculateTimeLeft = () => {
    const difference = +date - +new Date();
    let timeLeft = {};
    if (difference > 0) {
      timeLeft = {
        days: Math.floor(difference / (1000 * 60 * 60 * 24)),
        hours: Math.floor((difference / (1000 * 60 * 60)) % 24),
        minutes: Math.floor((difference / 1000 / 60) % 60),
        seconds: Math.floor((difference / 1000) % 60)
      };
    }
    return timeLeft;
  };

  const [timeLeft, setTimeLeft] = useState<any>(calculateTimeLeft());

  useEffect(() => {
    const timer = setTimeout(() => {
      setTimeLeft(calculateTimeLeft());
    }, 1000);
    return () => clearTimeout(timer);
  }, [calculateTimeLeft]);

  return (
    <>
      <StyledCountdown>
        <CountdownElement label="days">
          { timeLeft.days }
        </CountdownElement>
        <CountdownElement label="hours">
          { timeLeft.hours }
        </CountdownElement>
        <CountdownElement label="minutes">
          { timeLeft.minutes }
        </CountdownElement>
        <CountdownElement label="seconds">
          { timeLeft.seconds }
        </CountdownElement>
      </StyledCountdown>
      <p>Reward-Drop ETA date: { date.toUTCString() }</p>
    </>
  );
}