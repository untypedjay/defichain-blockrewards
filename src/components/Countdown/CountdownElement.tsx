import React from 'react';
import { StyledContent, StyledLabel } from '../../styles/CardStyle';

interface Props {
  children: number;
  label: string;
}

export default function CountdownElement({ children, label }: Props) {
  return (
    <div>
      <StyledContent>{ children }</StyledContent>
      <StyledLabel>{ label }</StyledLabel>
    </div>
  );
}