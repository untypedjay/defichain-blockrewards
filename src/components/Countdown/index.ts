export { default as Countdown } from './Countdown';
export { default as CountdownElement } from './CountdownElement';