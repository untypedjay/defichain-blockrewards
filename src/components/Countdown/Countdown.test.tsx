import ReactDOM from 'react-dom';
import React from 'react';
import Countdown from './Countdown';
import { cleanup, render } from '@testing-library/react';

afterEach(cleanup);

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Countdown date={new Date()}/>, div);
});

it('renders correctly', () => {
  render(<Countdown date={new Date()}/>);
});

it('matches snapshot', () => {
  const { asFragment } = render(<Countdown date={new Date()}/>);
  expect(asFragment()).toMatchSnapshot();
});