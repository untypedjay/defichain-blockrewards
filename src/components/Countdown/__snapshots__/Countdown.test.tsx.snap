// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-dlfnbm TzpTa"
  >
    <div>
      <p
        class="sc-bdfBwQ dHtfKI"
      />
      <p
        class="sc-gsTCUz fEJTeQ"
      >
        days
      </p>
    </div>
    <div>
      <p
        class="sc-bdfBwQ dHtfKI"
      />
      <p
        class="sc-gsTCUz fEJTeQ"
      >
        hours
      </p>
    </div>
    <div>
      <p
        class="sc-bdfBwQ dHtfKI"
      />
      <p
        class="sc-gsTCUz fEJTeQ"
      >
        minutes
      </p>
    </div>
    <div>
      <p
        class="sc-bdfBwQ dHtfKI"
      />
      <p
        class="sc-gsTCUz fEJTeQ"
      >
        seconds
      </p>
    </div>
  </div>
  <p>
    Reward-Drop ETA date: Sun, 28 Mar 2021 21:47:06 GMT
  </p>
</DocumentFragment>
`;
